diff --git a/default.cfg.example b/default.cfg.example
index db1f768..e97d1bc 100644
--- a/default.cfg.example
+++ b/default.cfg.example
@@ -15,6 +15,9 @@ all_currencies = STR,BTC,BTS,CLAM,DOGE,DASH,LTC,MAID,XMR,XRP,ETH,FCT,#BTG
 all_currencies = USD,BTC,BCH,ETH,XRP,IOT,XMR,LTC,OMG,ETC,EOS,DSH,ZEC,#BTG
 
 [BOT]
+#Debug mode, set to True to enable API related verbose debug messages in the console
+api_debug_log = False
+
 #Custom name of the bot, that will be displayed in html page
 label = Lending Bot
 
@@ -51,11 +54,11 @@ gaptop = 200
 #If set to 0 all offers will be placed for a 2 day period (0.003-5)
 # Poloniex max lending period: 60 days
 # Bitfinex max lending period: 120 days
-xdaythreshold = 0.2
-xdays = 60
-#When xdayspread is set, lending days will be incremented linear from 2 days at (xdaythreshold/xdayspread) lent rate
-#to (xdays) days at (xdaythreshold) lent rate. (0-10)
-#xdayspread = 2
+xdaythreshold = 0.050:20,0.058:30,0.060:45,0.063:60,0.070:120
+# These configs are not used any more
+#xdays = 30
+#xdayspread = 1.2
+
 
 #Auto-transfer of funds from exchange to lending balance.
 #Enter ALL to transfer all coins, enter ACTIVE to transfer any coins you have in your lending account when the bot starts, these can be mixed.
@@ -95,8 +98,10 @@ hideCoins = True
 #instead of mindailyrate. This only works on bitfinex.
 #frrasmin = False
 
-# The Flash Return Rate Delta (frrdelta) is a dynamic funding rate option that allows you to specify an offset from the Flash Return Rate. This only works on bitfinex and ffrasmin = TRUE
-frrdelta = 0.0000
+# The Flash Return Rate Delta min/max (frrdelta_min/max) are dynamic funding rate options that allow you to specify offset range from the Flash Return Rate. This only works on bitfinex and frrasmin = TRUE
+# 这个值要乘以 100 才是App和网站显示的百分比的值，负值则是基于frr再减去个delta
+frrdelta_min = 0.00000
+frrdelta_max = 0.00008
 
 #This option creates a json log file instead of console output which includes the most recent status.
 #Uncomment both jsonfile and jsonlogsize to enable.
@@ -145,7 +150,8 @@ frrdelta = 0.0000
 #gapbottom = 20
 #gaptop = 400
 #frrasmin = False
-#frrdelta = 0.0000
+#frrdelta_min = 0.0000
+#frrdelta_max = 0.0008
 
 #[CLAM]
 #minloansize = 1
@@ -158,7 +164,9 @@ frrdelta = 0.0000
 #gapbottom = 10
 #gaptop = 20
 #frrasmin = False
-#frrdelta = 0.0000
+#frrdelta_min = 0.0000
+#frrdelta_max = 0.0008
+
 
 [notifications]
 notify_new_loans = False
diff --git a/docs/configuration.rst b/docs/configuration.rst
index 419a20a..dcb572b 100644
--- a/docs/configuration.rst
+++ b/docs/configuration.rst
@@ -104,9 +104,9 @@ Min and Max Rates
     - This will only be used if the frr is above your ``mindailyrate``. So which ever is highest at the time of the loan will be used.
     - This options only works on Bitfinex.
 
-- ``frrdelta`` tells the bot whether or not to use the `flash return rate <https://support.bitfinex.com/hc/en-us/articles/115003284729-What-is-the-FRR-Delta->`
-    - Default value: 0.0000
-    - Allowed range: 0.0000 -/+ 7   
+- ``frrdelta_min`` and ``frrdelta_max`` are the delta range based on the `flash return rate <https://support.bitfinex.com/hc/en-us/articles/115003284729-What-is-the-FRR-Delta->`
+    - Default value: 0.0000/0.00008
+    - Allowed range: 
     - This will only be used if the frr is above your ``mindailyrate``. So which ever is highest at the time of the loan will be used.
     - This options only works on Bitfinex.
 
@@ -302,7 +302,8 @@ Configuration should look like this::
     gapbottom = 10
     gaptop = 20
     frrasmin = true
-    frrdelta = 0.000000
+    frrdelta_min = 0.00000
+    frrdelta_max = 0.00008
 
 
 Advanced logging and Web Display
diff --git a/lendingbot.py b/lendingbot.py
index 179b795..61459e3 100755
--- a/lendingbot.py
+++ b/lendingbot.py
@@ -108,11 +108,21 @@ try:
         try:
             dns_cache = {}  # Flush DNS Cache
             Data.update_conversion_rates(output_currency, json_output_enabled)
-            PluginsManager.before_lending()
-            Lending.transfer_balances()
-            Lending.cancel_all()
-            Lending.lend_all()
-            PluginsManager.after_lending()
+
+            if Lending.lending_paused != Lending.last_lending_status:
+                if not Lending.lending_paused:
+                    log.log('Lending running')
+                else:
+                    log.log('Lending paused')
+                Lending.last_lending_status = Lending.lending_paused
+
+            if not Lending.lending_paused:
+                PluginsManager.before_lending()
+                Lending.transfer_balances()
+                Lending.cancel_all()
+                Lending.lend_all()
+                PluginsManager.after_lending()
+
             log.refreshStatus(Data.stringify_total_lent(*Data.get_total_lent()),
                               Data.get_max_duration(end_date, "status"))
             log.persistStatus()
diff --git a/modules/Bitfinex.py b/modules/Bitfinex.py
index 9ec0a10..3af9122 100644
--- a/modules/Bitfinex.py
+++ b/modules/Bitfinex.py
@@ -30,7 +30,7 @@ class Bitfinex(ExchangeApi):
         self.symbols = []
         self.ticker = {}
         self.tickerTime = 0
-	self.baseCurrencies = ['USD', 'BTC', 'ETH']
+        self.baseCurrencies = ['USD', 'BTC', 'ETH']
         self.all_currencies = self.cfg.get_all_currencies()
         self.usedCurrencies = []
         self.timeout = int(self.cfg.get("BOT", "timeout", 30, 1, 180))
@@ -71,6 +71,10 @@ class Bitfinex(ExchangeApi):
             "Connection": "close"
         }
 
+    def debug_log(self, msg):
+        if self.api_debug_log:
+            self.log.log(msg)
+
     def _request(self, method, request, payload=None, verify=True):
         try:
 
@@ -78,8 +82,10 @@ class Bitfinex(ExchangeApi):
             url = '{}{}'.format(self.url, request)
             if method == 'get':
                 r = requests.get(url, timeout=self.timeout, headers={'Connection': 'close'})
+                self.debug_log("GET: {}".format(url))
             else:
                 r = requests.post(url, headers=payload, verify=verify, timeout=self.timeout)
+                self.debug_log("POST: {} headers={}".format(url, payload))
 
             if r.status_code != 200:
                 if r.status_code == 502 or r.status_code in range(520, 527, 1):
@@ -91,6 +97,7 @@ class Bitfinex(ExchangeApi):
 
             # Check in case something has gone wrong and the timer is too big
             self.reset_request_timer()
+            self.debug_log("Response: {}".format(r.text))
             return r.json()
 
         except Exception as ex:
diff --git a/modules/Configuration.py b/modules/Configuration.py
index d321cc9..7c2e332 100755
--- a/modules/Configuration.py
+++ b/modules/Configuration.py
@@ -106,7 +106,8 @@ def get_coin_cfg():
                 coin_cfg[cur]['gapbottom'] = Decimal(get(cur, 'gapbottom', False, 0))
                 coin_cfg[cur]['gaptop'] = Decimal(get(cur, 'gaptop', False, coin_cfg[cur]['gapbottom']))
                 coin_cfg[cur]['frrasmin'] = getboolean(cur, 'frrasmin', getboolean('BOT', 'frrasmin'))
-                coin_cfg[cur]['frrdelta'] = Decimal(get(cur, 'frrdelta', 0.0000))
+                coin_cfg[cur]['frrdelta_min'] = Decimal(get(cur, 'frrdelta_min', 0.0000))
+                coin_cfg[cur]['frrdelta_max'] = Decimal(get(cur, 'frrdelta_max', 0.00008))
 
             except Exception as ex:
                 ex.message = ex.message if ex.message else str(ex)
diff --git a/modules/Data.py b/modules/Data.py
index 935fd9c..3e92c3c 100644
--- a/modules/Data.py
+++ b/modules/Data.py
@@ -109,18 +109,21 @@ def update_conversion_rates(output_currency, json_output_enabled):
                     log.updateOutputCurrency('currency', output_currency)
         if not output_currency_found:  # fetch output currency rate from blockchain.info
             url = "https://blockchain.info/tobtc?currency={0}&value=1".format(output_currency)
+        try:
+            response = urlopen(url).read()
             try:
-                highest_bid = json.loads(urlopen(url).read())
+                highest_bid = json.loads(response)
                 log.updateOutputCurrency('highestBid', 1 / float(highest_bid))
                 log.updateOutputCurrency('currency', output_currency)
             except ValueError:
-                log.log_error("Failed to find the exchange rate for outputCurrency {0}! Using BTC as output currency"
-                              .format(output_currency))
-                log.log_error("Make sure that {0} is either traded on the exchange or supported by blockchain.info: {1}"
-                              .format(output_currency, "https://blockchain.info/api/exchange_rates_api"))
-            except Exception:
-                log.log_error("Can't connect to {0} using BTC as the output currency".format(url))
-
+                try:
+                    highest_bid = response.decode('utf-8')
+                    log.updateOutputCurrency('highestBid', 1 / float(highest_bid))
+                    log.updateOutputCurrency('currency', output_currency)
+                except ValueError:
+                    log.log_error("Failed to decode response as plain text or JSON")
+        except Exception:
+            log.log_error("Can't connect to {0} using BTC as the output currency".format(url))
 
 def get_lending_currencies():
     currencies = []
diff --git a/modules/Lending.py b/modules/Lending.py
index 3176bcc..3bd395b 100644
--- a/modules/Lending.py
+++ b/modules/Lending.py
@@ -20,9 +20,7 @@ max_daily_rate = 0
 spread_lend = 0
 gap_bottom_default = 0
 gap_top_default = 0
-xday_threshold = 0
-xday_spread = 0
-xdays = 0
+xday_threshold = ""
 min_loan_size = 0
 min_loan_sizes = {}
 end_date = None
@@ -40,11 +38,24 @@ gap_mode_default = ""
 scheduler = None
 exchange = None
 frrasmin = False
-frrdelta = 0.0
+
+frrdelta_cur_step = 0
+frrdelta_min = 0.0
+frrdelta_max = 0.0
+debug_on = False
+lending_paused = False
+last_lending_status = None
 
 # limit of orders to request
 loanOrdersRequestLimit = {}
-defaultLoanOrdersRequestLimit = 100
+defaultLoanOrdersRequestLimit = 5
+# FIXME: Make this configurable
+compete_rate = 0.00064
+
+
+def debug_log(msg):
+    if debug_on:
+        print('DEBUG: ' + msg)
 
 
 def init(cfg, api1, log1, data, maxtolend, dry_run1, analysis, notify_conf1):
@@ -58,9 +69,9 @@ def init(cfg, api1, log1, data, maxtolend, dry_run1, analysis, notify_conf1):
     notify_conf = notify_conf1
 
     global sleep_time, sleep_time_active, sleep_time_inactive, min_daily_rate, max_daily_rate, spread_lend, \
-        gap_bottom_default, gap_top_default, xday_threshold, xday_spread, xdays, min_loan_size, end_date, coin_cfg, \
+        gap_bottom_default, gap_top_default, xday_threshold, min_loan_size, end_date, coin_cfg, \
         min_loan_sizes, dry_run, transferable_currencies, keep_stuck_orders, hide_coins, scheduler, gap_mode_default, \
-        exchange, analysis_method, currencies_to_analyse, all_currencies, frrasmin, frrdelta
+        exchange, analysis_method, currencies_to_analyse, all_currencies, frrasmin, frrdelta_min, frrdelta_max
 
     exchange = Config.get_exchange()
 
@@ -73,10 +84,8 @@ def init(cfg, api1, log1, data, maxtolend, dry_run1, analysis, notify_conf1):
     gap_mode_default = Config.get_gap_mode("BOT", "gapMode")
     gap_bottom_default = Decimal(Config.get("BOT", "gapbottom", None, 0))
     gap_top_default = Decimal(Config.get("BOT", "gaptop", None, gap_bottom_default))
-    xday_threshold = float(Config.get("BOT", "xdaythreshold", None, 0.003, 5)) / 100
-    xday_spread = float(Config.get('BOT', 'xdayspread', 0, 0, 10))
+    xday_threshold = Config.get("BOT", "xdaythreshold")
     maxPeriod = 120 if exchange == 'BITFINEX' else 60
-    xdays = str(Config.get("BOT", "xdays", None, 2, maxPeriod))
     min_loan_size = Decimal(Config.get("BOT", 'minloansize', None, 0.01))
     end_date = Config.get('BOT', 'endDate')
     coin_cfg = Config.get_coin_cfg()
@@ -88,7 +97,8 @@ def init(cfg, api1, log1, data, maxtolend, dry_run1, analysis, notify_conf1):
     keep_stuck_orders = Config.getboolean('BOT', "keepstuckorders", True)
     hide_coins = Config.getboolean('BOT', 'hideCoins', True)
     frrasmin = Config.getboolean('BOT', 'frrasmin', False)
-    frrdelta = Decimal(Config.get('BOT', 'frrdelta', 0.0000))
+    frrdelta_min = Decimal(Config.get('BOT', 'frrdelta_min', 0.0000))
+    frrdelta_max = Decimal(Config.get('BOT', 'frrdelta_max', 0.00008))
     analysis_method = Config.get('Daily_min', 'method', 'percentile')
     if analysis_method not in ['percentile', 'MACD']:
         raise ValueError("analysis_method: \"{0}\" is not valid, must be percentile or MACD".format(analysis_method))
@@ -163,20 +173,39 @@ def get_min_loan_size(currency):
         return min_loan_size
     return Decimal(min_loan_sizes[currency])
 
-
-def create_lend_offer(currency, amt, rate):
-    days = '2'
+# parse config like "0.050:25,0.058:30,0.060:45,0.064:60,0.070:120", i.e. rate:days pairs,
+# and return the rates, days list
+def parse_xday_threshold(xday_threshold):
+    rates = []
+    xdays = []
+    if xday_threshold:
+        for pair in xday_threshold.split(','):
+            rate, day = pair.split(':')
+            rates.append(float(rate) / 100)
+            xdays.append(day)
+    return rates, xdays
+
+def create_lend_offer(currency, amt, rate, days='2'):
     if float(rate) > 0.0001:
         rate = float(rate) - 0.000001  # lend offer just bellow the competing one
     amt = "%.8f" % Decimal(amt)
-    if xday_threshold > 0:
-        if float(rate) >= xday_threshold:
-            days = xdays
-        elif xday_spread and xday_spread > 0:
-            xday_threshold_min = xday_threshold / xday_spread
-            if float(rate) > xday_threshold_min:
-                m = (float(xdays) - 2) / (xday_threshold - xday_threshold_min)
-                days = str(int(round(m * (float(rate) - xday_threshold_min) + 2)))
+    rates, xdays = parse_xday_threshold(xday_threshold)
+    if days == '2' and len(rates) > 0:
+        # map rate to xdays, use intepolation if rate is not in the list
+        if float(rate) < rates[0]:
+            days = xdays[0]
+        else:
+            for i in range(len(rates)):
+                if float(rate) <= rates[i]:
+                    # linear interpolation
+                    days = str(int(xdays[i - 1]) + int((int(xdays[i]) - int(xdays[i - 1])) * (float(rate) - rates[i - 1]) /
+                        (rates[i] - rates[i - 1])))
+                    break
+            else:
+                # If rate is greater than the last rate, use the last xdays
+                days = xdays[-1]
+        print "Using xday threshold: rate={0}, days={1}".format(rate, days)
+
     if Config.has_option('BOT', 'endDate'):
         days_remaining = int(Data.get_max_duration(end_date, "order"))
         if int(days_remaining) <= 2:
@@ -190,7 +219,7 @@ def create_lend_offer(currency, amt, rate):
             days = str(days_remaining)
     if not dry_run:
         msg = api.create_loan_offer(currency, amt, days, 0, rate)
-        if days == xdays and notify_conf['notify_xday_threshold']:
+        if days == xdays[-1] and notify_conf['notify_xday_threshold']:
             text = "{0} {1} loan placed for {2} days at a rate of {3:.4f}%".format(amt, currency, days, rate * 100)
             log.notify(text, notify_conf)
         log.offer(amt, currency, rate, days, msg)
@@ -200,6 +229,8 @@ def cancel_all():
     loan_offers = api.return_open_loan_offers()
     available_balances = api.return_available_account_balances('lending')
     for CUR in loan_offers:
+        if CUR not in all_currencies:
+            continue
         if CUR in coin_cfg and coin_cfg[CUR]['maxactive'] == 0:
             # don't cancel disabled coin
             continue
@@ -246,6 +277,8 @@ def lend_all():
             if "rawbtc" in cur1:
                 ticker = api.return_ticker()
             break
+    log.log("Lending balances: " + str(lending_balances))
+    log.log("All currencies: " + str(all_currencies))
     try:
         for cur in lending_balances:
             if cur in all_currencies:
@@ -263,21 +296,42 @@ def get_frr_or_min_daily_rate(cur):
     :param cur: The currency which to check
     :return: The better of the two rates (FRR and min daily rate)
     """
+    global frrdelta_cur_step
+    global frrdelta_min, frrdelta_max
     if cur in coin_cfg:
         min_daily_rate = Decimal(coin_cfg[cur]['minrate'])
         frrasmin = coin_cfg[cur]['frrasmin']
-        frrdelta = Decimal(coin_cfg[cur]['frrdelta']) / 100
+        frrdelta_min = Decimal(coin_cfg[cur]['frrdelta_min']) / 100
+        frrdelta_max = Decimal(coin_cfg[cur]['frrdelta_max']) / 100
     else:
         min_daily_rate = Decimal(Config.get("BOT", "mindailyrate", None, 0.003, 5)) / 100
         frrasmin = Config.getboolean('BOT', 'frrasmin', False)
-        frrdelta = Decimal(Config.get('BOT', 'frrdelta', 0.0000))
+
+    if frrdelta_min > frrdelta_max:
+        frrdelta_min, frrdelta_max = frrdelta_max, frrdelta_min
+    # Let's use hard coded steps for now
+    frrdelta_steps = 5
+    frrdelta_step = (frrdelta_max - frrdelta_min) / frrdelta_steps
+
+    if frrdelta_cur_step > frrdelta_steps:
+        frrdelta_cur_step = 0
+    frrdelta = frrdelta_min + (frrdelta_step * frrdelta_cur_step)
+    frrdelta_cur_step += 1
+
+    log.log("Using frrasmin {0} for {1}".format(frrasmin, cur))
+    log.log("Using frrdelta {0}% + {1}% = {2}% for {3}".format(
+            frrdelta_min * 100,
+            (frrdelta - frrdelta_min) * 100,
+            frrdelta * 100,
+            cur))
 
     if exchange == 'BITFINEX' and frrasmin:
         frr_rate = Decimal(api.get_frr(cur)) + frrdelta
         if frr_rate > min_daily_rate:
-            log.log("Using FRR as mindailyrate {0}% for {1}".format(frr_rate * 100, cur))
+            log.log("Using FRR as mindailyrate {:.6f}% for {}".format(frr_rate * 100, cur))
             return frr_rate
 
+    log.log("Using min_daily_rate {:.6f}% for {}".format(min_daily_rate * 100, cur))
     return min_daily_rate
 
 
@@ -293,14 +347,15 @@ def get_min_daily_rate(cur):
             coin_cfg_alerted[cur] = True
             log.log('Using custom mindailyrate ' + str(cur_min_daily_rate * 100) + '% for ' + cur)
     if Analysis and cur in currencies_to_analyse:
+        # TODO: 这里看下建议的 rate 是怎么来的？
         recommended_min = Analysis.get_rate_suggestion(cur, method=analysis_method)
         if cur_min_daily_rate < recommended_min:
-            log.log("Using {0} as mindailyrate {1}% for {2}".format(analysis_method, recommended_min * 100, cur))
-            cur_min_daily_rate = recommended_min
+            log.log("Suggest to use {0} as mindailyrate {1}% for {2}".format(analysis_method, recommended_min * 100, cur))
+            #cur_min_daily_rate = recommended_min
     return Decimal(cur_min_daily_rate)
 
 
-def construct_order_book(active_cur):
+def construct_order_books(active_cur):
     # make sure we have a request limit for this currency
     if active_cur not in loanOrdersRequestLimit:
         loanOrdersRequestLimit[active_cur] = defaultLoanOrdersRequestLimit
@@ -309,15 +364,40 @@ def construct_order_book(active_cur):
     if len(loans) == 0:
         return False
 
-    rate_book = []
-    volume_book = []
-    for offer in loans['offers']:
-        rate_book.append(offer['rate'])
-        volume_book.append(offer['amount'])
-    return {'rates': rate_book, 'volumes': volume_book}
+    resps = []
+    for load_type in ('demands', 'offers'):
+        rate_book = []
+        volume_book = []
+        rangeMax_book = []
+        for load in loans[load_type]:
+            rate_book.append(load['rate'])
+            volume_book.append(load['amount'])
+            rangeMax_book.append(load['rangeMax'])
+        resp = {'rates': rate_book, 'volumes': volume_book, 'rangeMax': rangeMax_book}
+        debug_log("construct %s: %s" % (load_type, str(resp)))
+        resps.append(resp)
+
+    return resps
 
 
 def get_gap_rate(active_cur, gap, order_book, cur_total_balance, raw=False):
+    """
+    Calculates the gap rate for a given active currency.
+
+    Args:
+        active_cur (str): The active currency.
+        gap (float): The gap value.
+        order_book (dict): The order book containing volumes and rates.
+        cur_total_balance (Decimal): The current total balance.
+        raw (bool, optional): Whether to use the raw gap value. Defaults to False.
+
+    Returns:
+        float: The calculated gap rate.
+
+    Raises:
+        StopIteration: If there are not enough offers in the response.
+
+    """
     if raw:
         gap_expected = gap
     else:
@@ -326,7 +406,7 @@ def get_gap_rate(active_cur, gap, order_book, cur_total_balance, raw=False):
     i = 0
     while gap_sum < gap_expected:
         if i == len(order_book['volumes']) - 1 and len(order_book['volumes']) == loanOrdersRequestLimit[active_cur]:
-            loanOrdersRequestLimit[active_cur] += defaultLoanOrdersRequestLimit
+            #loanOrdersRequestLimit[active_cur] += defaultLoanOrdersRequestLimit
             log.log(active_cur + ': Not enough offers in response, adjusting request limit to ' + str(
                 loanOrdersRequestLimit[active_cur]))
             raise StopIteration
@@ -347,11 +427,14 @@ def get_cur_spread(spread, cur_active_bal, active_cur):
 
 def construct_orders(cur, cur_active_bal, cur_total_balance, ticker):
     cur_spread = get_cur_spread(spread_lend, cur_active_bal, cur)
-    top_rate, bottom_rate = get_gap_mode_rates(cur, cur_active_bal, cur_total_balance, ticker)
-    gap_diff = top_rate - bottom_rate
     if cur_spread == 1:
+        print('skip get_gap_mode_rates ...')
         rate_step = 0
+        bottom_rate = 0
     else:
+        print('call get_gap_mode_rates ...')
+        top_rate, bottom_rate = get_gap_mode_rates(cur, cur_active_bal, cur_total_balance, ticker)
+        gap_diff = top_rate - bottom_rate
         rate_step = gap_diff / (cur_spread - 1)
 
     order_rates = []
@@ -375,14 +458,16 @@ def construct_orders(cur, cur_active_bal, cur_total_balance, ticker):
     remainder = cur_active_bal - sum(new_order_amounts)
     if remainder > 0:  # If truncating causes remainder, add that to first order.
         new_order_amounts[0] += remainder
-    return {'amounts': new_order_amounts, 'rates': new_order_rates}
+    resp = {'amounts': new_order_amounts, 'rates': new_order_rates}
+    debug_log('Constructing orders: ' + str(resp))
+    return resp
 
 
 def get_gap_mode_rates(cur, cur_active_bal, cur_total_balance, ticker):
     global gap_mode_default, gap_bottom_default, gap_top_default  # To be able to change them later if needed.
     gap_mode, gap_bottom, gap_top = gap_mode_default, gap_bottom_default, gap_top_default
     use_gap_cfg = False
-    order_book = construct_order_book(cur)
+    _, order_book = construct_order_books(cur)
     if cur in coin_cfg:  # Get custom values specific to coin
         cfg = coin_cfg[cur]
         if cfg.get('gapmode', False) and cfg.get('gapbottom', False) and cfg.get('gaptop', False):
@@ -420,6 +505,7 @@ def get_gap_mode_rates(cur, cur_active_bal, cur_total_balance, ticker):
             gap_bottom_default = 10
             gap_top_default = 200
         return get_gap_mode_rates(cur, cur_active_bal, cur_total_balance, ticker)  # Start over with new defaults
+    debug_log("gap_mode: " + gap_mode + ", top_rate: " + str(top_rate) + ", bottom_rate: " + str(bottom_rate))
     return [Decimal(top_rate), Decimal(bottom_rate)]
 
 
@@ -434,7 +520,8 @@ def lend_cur(active_cur, total_lent, lending_balances, ticker):
 
     # log total coin
     log.updateStatusValue(active_cur, "totalCoins", (Decimal(active_cur_total_balance)))
-    order_book = construct_order_book(active_cur)
+    demand_book, order_book = construct_order_books(active_cur)
+
     if not order_book or len(order_book['rates']) == 0 or not cur_min_daily_rate:
         return 0
 
@@ -460,8 +547,15 @@ def lend_cur(active_cur, total_lent, lending_balances, ticker):
         else:
             rate = orders['rates'][i]
 
+        days = '2'
+        # Check demand_book for competing offers
+        if demand_book and float(demand_book['rates'][0]) > compete_rate:
+            rate = demand_book['rates'][0]
+            days = str(demand_book['rangeMax'][0])
+            log.log("Competing offer found for " + active_cur + " at " + str(float(rate) * 100) + "% for " + days + " days.")
+
         try:
-            create_lend_offer(active_cur, orders['amounts'][i], rate)
+            create_lend_offer(active_cur, orders['amounts'][i], rate, days)
         except Exception as msg:
             if "Amount must be at least " in str(msg):
                 import re
diff --git a/modules/WebServer.py b/modules/WebServer.py
index 0a326de..d2a1567 100644
--- a/modules/WebServer.py
+++ b/modules/WebServer.py
@@ -1,6 +1,11 @@
 # coding=utf-8
 import threading
 import os
+import json
+from decimal import Decimal
+from decimal import InvalidOperation
+
+import modules.Lending as Lending
 
 server = None
 web_server_ip = "0.0.0.0"
@@ -71,6 +76,50 @@ def start_web_server():
                     return None
                 return SimpleHTTPServer.SimpleHTTPRequestHandler.send_head(self)
 
+            def do_GET(self):
+                if self.path == '/pause_lending':
+                    Lending.lending_paused = True
+                    self.send_response(200)
+                    self.end_headers()
+                    self.wfile.write(b'Lending paused')
+                elif self.path == '/resume_lending':
+                    Lending.lending_paused = False
+                    self.send_response(200)
+                    self.end_headers()
+                    self.wfile.write(b'Lending resumed')
+                elif self.path == '/get_status':
+                    self.send_response(200)
+                    self.send_header('Content-Type', 'application/json')
+                    self.end_headers()
+                    self.wfile.write(json.dumps({"lending_paused": Lending.lending_paused}))
+                else:
+                    return SimpleHTTPServer.SimpleHTTPRequestHandler.do_GET(self)
+
+            def do_POST(self):
+                if self.path == '/set_config':
+                    content_length = int(self.headers['Content-Length'])  # 获取数据长度
+                    post_data = self.rfile.read(content_length)  # 读取POST数据
+                    config_data = json.loads(post_data)
+
+                    # 更新配置值
+                    if 'frrdelta_min' in config_data and 'frrdelta_max' in config_data:
+                        try:
+                            Lending.frrdelta_min = Decimal(config_data['frrdelta_min'])
+                            Lending.frrdelta_max = Decimal(config_data['frrdelta_max'])
+                            response = {"success": True, "frrdelta_min": str(Lending.frrdelta_min), "frrdelta_max": str(Lending.frrdelta_max)}
+                        except (ValueError, TypeError, InvalidOperation) as e:
+                            response = {"success": False, "error": str(e)}
+                    else:
+                        response = {"success": False, "error": "Invalid configuration key"}
+
+                    # 发送响应
+                    self.send_response(200 if response["success"] else 400)
+                    self.send_header('Content-Type', 'application/json')
+                    self.end_headers()
+                    self.wfile.write(json.dumps(response))
+                else:
+                    self.send_error(404, "File not found")
+
         global server
         SocketServer.TCPServer.allow_reuse_address = True
         server = SocketServer.TCPServer((host, port), QuietHandler)
diff --git a/www/lendingbot.html b/www/lendingbot.html
index c672549..c3f5c98 100644
--- a/www/lendingbot.html
+++ b/www/lendingbot.html
@@ -37,6 +37,9 @@
     <script src="https://cdnjs.cloudflare.com/ajax/libs/toastr.js/latest/toastr.min.js"></script>
     <link href="https://cdnjs.cloudflare.com/ajax/libs/toastr.js/latest/toastr.min.css" rel="stylesheet">
 
+    <!-- For some icons -->
+    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0-beta3/css/all.min.css">
+
     <script src="lendingbot.js?version=2" ></script>
 
     <style>
@@ -116,6 +119,11 @@
                             Settings
                         </a>
                     </li>
+                    <li data-toggle="collapse" data-target=".navbar-collapse.in">
+                        <button type="button" class="btn btn-primary navbar-btn" id="pauseButton">
+                            <i class="fas fa-pause"></i> Pause Lending
+                        </button>
+                    </li>
                 </ul>
             </div>
         </div>
@@ -222,6 +230,15 @@
                             </label>
                         </div>
 
+                        <div class="form-group">
+                            <label for="frrdelta_min">FRR minimum delta value (+-0.003 etc.)</label>
+                            <input type="number" class="form-control" id="frrdelta_min" placeholder="(-0.00009 ~ 0.00009)">
+                        </div>
+
+                        <div class="form-group">
+                            <label for="frrdelta_max">FRR maximum delta value (+-0.003 etc.)</label>
+                            <input type="number" class="form-control" id="frrdelta_max" placeholder="(-0.00009 ~ 0.00009)">
+                        </div>
                     </form>
                 </div>
                 <div class="modal-footer">
diff --git a/www/lendingbot.js b/www/lendingbot.js
index 6b8919c..dd801b3 100644
--- a/www/lendingbot.js
+++ b/www/lendingbot.js
@@ -392,6 +392,14 @@ function loadSettings() {
     timespanNames.forEach(function(t) {
         $('input[data-timespan="' + t + '"]').prop('checked', true);
     });
+
+    // frrdelta_min
+    var frrdelta_min = localStorage.getItem('frrdelta_min') || 0.00001;
+    $('#frrdelta_min').val(frrdelta_min);
+
+    // frrdelta_max
+    var frrdelta_max = localStorage.getItem('frrdelta_max') || 0.00001;
+    $('#frrdelta_max').val(frrdelta_max);
 }
 
 function doSave() {
@@ -432,6 +440,25 @@ function doSave() {
         effRateMode = localStorage.effRateMode;
     }
 
+    var frrdelta_min = parseFloat($('#frrdelta_min').val());
+    if(frrdelta_min < -0.003 || frrdelta_min > 0.003) {
+        alert('Please input a value between -0.003 and 0.003 for frrdelta_min');
+        return false;
+    }
+    localStorage.setItem('frrdelta_min', frrdelta_min.toString());
+
+    var frrdelta_max = parseFloat($('#frrdelta_max').val());
+    if(frrdelta_max < frrdelta_min || frrdelta_max > 0.003) {
+        alert('Please input a value between frrdelta_min and 0.003 for frrdelta_max');
+        return false;
+    }
+    localStorage.setItem('frrdelta_max', frrdelta_max.toString());
+
+    setConfig({
+        "frrdelta_min": frrdelta_min,
+        "frrdelta_max": frrdelta_max
+    });
+
     toastr.success("Settings saved!");
     $('#settings_modal').modal('hide');
 
@@ -460,6 +487,42 @@ function bsNavbarBugWorkaround() {
     });
 }
 
+function updateButtonStatus(isPaused) {
+    var buttonHtml = isPaused ? '<i class="fas fa-play"></i> Resume Lending' : '<i class="fas fa-pause"></i> Pause Lending';
+    $('#pauseButton').html(buttonHtml);
+}
+
+function handle_pause_button() {
+    // 查询当前的状态
+    $.get('/get_status', function(data) {
+        updateButtonStatus(data.lending_paused);
+    });
+
+    $('#pauseButton').click(function() {
+        var isPaused = $('#pauseButton').text().trim() == 'Resume Lending';
+        var url = isPaused ? '/resume_lending' : '/pause_lending';
+
+        $.get(url, function() {
+            updateButtonStatus(!isPaused);
+        });
+    });
+}
+
+function setConfig(config) {
+    $.ajax({
+        url: '/set_config',
+        type: 'POST',
+        contentType: 'application/json',
+        data: JSON.stringify(config),
+        success: function(response) {
+            console.log(response);
+        },
+        error: function(xhr, status, error) {
+            console.error("Error setting config:", error);
+        }
+    });
+}
+
 $(document).ready(function () {
     toastr.options = {
         "positionClass": "toast-top-center"
@@ -467,4 +530,5 @@ $(document).ready(function () {
 
     update();
     bsNavbarBugWorkaround();
+    handle_pause_button();
 });
